### PR TITLE
Add HBase datasource for nebula exchange

### PR DESCRIPTION
**Pull request by Nicole00**
Import data from HBase to nebula, just configure the src/main/resources/application.conf.

### CODE REVIEW

File: tools/exchange/pom.xml [hunk 1]
```diff
@@ -636,6 +641,31 @@
...
+            <version>${hbase.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hbase</groupId>
+            <artifactId>hbase-common</artifactId>
+            <version>${hbase.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hbase</groupId>
+            <artifactId>hbase-server</artifactId>
```
Comment (darionyaphet) [inline tools/exchange/pom.xml:661]: why import `hbase-server`?

Comment (Nicole00) [reply tools/exchange/pom.xml:661]: hbase-server is useless, i will remove it, thanks.

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala [hunk 2]
```diff
@@ -190,3 +198,60 @@ class NebulaReader(override val session: SparkSession, nebulaConfig: ServerDataS
...
+      }
+      Row.fromSeq(values.toList)
+    })
+    val schema = StructType(
+      fields.map(field => DataTypes.createStructField(field, sourceSchema(field), true)))
+    val dataFrame = session.createDataFrame(values, schema)
+    dataFrame.printSchema()
+    LOG.info("hbase source data example: ")
+    dataFrame.show()
+    dataFrame
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala:255]: remove the debug info?

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      metaClient = new MetaClientImpl(hostPorts.toList.asJava)
+    } catch {
+      case e: TException => {
+        LOG.error("failed to get metaClient")
+        throw e
+      }
+    }
+    if (ErrorCode.SUCCEEDED != metaClient.connect()) {
+      LOG.error("meta client connect failed.")
+      sys.exit(-1)
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala:56]: Java virtual machine will stop ?

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  def getDataSourceFieldType(dataSourceConf: Config,
+                             nebulaConfig: Config): Map[String, DataType] = {
+    val nebulaFields = dataSourceConf.getStringList("nebula.fields")
+    val sourceFields = dataSourceConf.getStringList("fields")
+    val address      = nebulaConfig.getStringList("metaAddress")
+    val space        = nebulaConfig.getString("space")
+    val label        = dataSourceConf.getString("name")
+
+    val hostPorts: ListBuffer[HostAndPort] = new ListBuffer[HostAndPort]
+    for (i <- 0 until address.size()) {
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala:41]: if you can know the size of the List please specified it when initial

File: tools/exchange/src/main/resources/server_application.conf [hunk 5]
```diff
@@ -18,6 +18,8 @@
   # Nebula Graph relation config
   nebula: {
     addresses: ["127.0.0.1:3699"]
+    # If the data source is Hbase, metaAddress must be configured
+    metaAddress:["127.0.0.1:3699"]
```
Comment (darionyaphet) [inline tools/exchange/src/main/resources/server_application.conf:22]: `127.0.0.1:54500` is default address of the meta service?

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala [hunk 6]
```diff
@@ -192,3 +195,21 @@ case class PulsarSourceConfigEntry(override val category: SourceCategory.Value,
...
+  */
+case class HBaseSourceConfigEntry(override val category: SourceCategory.Value,
+                                  override val sentence: String,
+                                  host: String,
+                                  port: String,
+                                  table: String,
+                                  columnFamily: String,
+                                  fields: List[String],
+                                  sourceFieldSchema: Map[String, DataType])
+    extends ServerDataSourceConfigEntry() {
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala:211]: support more argument checker ?

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala [hunk 7]
```diff
@@ -539,6 +542,32 @@ object Configs {
...
+        if (config.hasPath("source.field")) {
+          fields.append(config.getString("source.field"))
+        }
+        if (config.hasPath("target.field")) {
+          fields.append(config.getString("target.field"))
+        }
+
+        HBaseSourceConfigEntry(
+          SourceCategory.HBASE,
+          getOrElse(config, "sentence", ""),
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala:563]: Actually, HBase do not support sentence?

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala (hunk 3 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala:56]: Yes, jvm will stop. Maybe it's better to throw exception.

File: tools/exchange/src/main/resources/server_application.conf (hunk 5 above)
Comment (Nicole00) [reply tools/exchange/src/main/resources/server_application.conf:22]: I forget the change the meta address's default port, it's 45500.

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala (hunk 6 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala:211]: Yeah,  argument check is required.

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala (hunk 7 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala:563]: Yes, sentence parameter for HBaseSourceConfigEntry should be removed.

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala (hunk 4 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/utils/NebulaUtils.scala:41]: I removed the loop and changed the code to :
val hostPorts = address.toArray.map(addr => HostAndPort.fromString(addr.toString)).toList

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala (hunk 2 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala]: OK

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/Exchange.scala [hunk 8]
```diff
@@ -79,6 +79,7 @@ object Exchange {
     val session = SparkSession
       .builder()
       .appName(PROGRAM_NAME)
+      .master("local")
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/Exchange.scala:82]: if we specified the master as local could we config it from command line?

Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/Exchange.scala:82]: Sorry, I forget to delete the test code.

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala [hunk 9]
```diff
@@ -539,6 +542,31 @@ object Configs {
                                 config.getString("service"),
                                 config.getString("admin"),
                                 options)
+      case SourceCategory.HBASE =>
+        val fields: ListBuffer[String] = new ListBuffer[String]
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala:546]: ditto

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala [hunk 10]
```diff
@@ -192,3 +196,26 @@ case class PulsarSourceConfigEntry(override val category: SourceCategory.Value,
...
+
+  require(host.trim.length != 0 && port.trim.length != 0 && NebulaUtils
+    .isNumic(port.trim) && table.trim.length > 0 && table.trim.length > 0 && columnFamily.trim.length > 0)
+
+  override val sentence: String = null
+
+  override def toString: String = {
+    s"HBase graph source, HBase host is $host, port is $port, table is $table"
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala:219]: HBase source ?

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala [hunk 11]
```diff
@@ -190,3 +198,57 @@ class NebulaReader(override val session: SparkSession, nebulaConfig: ServerDataS
...
+    val hbaseRDD: RDD[(ImmutableBytesWritable, Result)] = session.sparkContext.newAPIHadoopRDD(
+      scanConf,
+      classOf[TableInputFormat],
+      classOf[ImmutableBytesWritable],
+      classOf[Result])
+
+    val sourceSchema = hbaseConfig.sourceFieldSchema
+
+    val values = hbaseRDD.map(row => {
+      val values: ListBuffer[Any] = new ListBuffer[Any]
```
Comment (darionyaphet) [inline tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala:234]: `values` have shadowed maybe make reader confuse

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala (hunk 10 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/SourceConfigs.scala:219]: remove graph

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala (hunk 11 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/reader/ServerBaseReader.scala:234]: changed the first `values` to `rowRDD`

File: tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala (hunk 9 above)
Comment (Nicole00) [reply tools/exchange/src/main/scala/com/vesoft/nebula/tools/importer/config/Configs.scala:546]: the loop has been changed to 
`fields.append(config.getStringList("fields").asScala.map(field => field):_*)`

**Review by darionyaphet — APPROVED** [review]

**Pull request merged**